### ISSUE TITLE
Code-first migrations with one-to-zero-or-one does not generate the foreign key constraint for both entities

### ISSUE THREAD

**Issue by IRCraziestTaxi** [dotnet/efcore#16520]
### Steps to reproduce
We have multiple entities (e.g. `Document`, `Media`, etc.) that reference a `Change` entity. Each entity has a `Change` with a `[ForeignKey]` attribute which references the `ChangeId` property. The `Change` entity has an optional reference to each entity that can reference it (the `*Id` properties are `int?` rather than `int`).

The problem is that, with code-first migrations, only one side of the relationship has the foreign key constraint enforced. (Will elaborate after code snippets.)

Here are the entities and relevant properties:

```c#
public class Change
{
        [Key]
        public int ChangeId { get; set; }

        [ForeignKey(nameof(DocumentId))]
        public Document Document { get; set; }

        public int? DocumentId { get; set; }

        [ForeignKey(nameof(MediaId))]
        public Media Media { get; set; }

        public int? MediaId { get; set; }
}

public class Document
{
        [ForeignKey(nameof(ChangeId))]
        public Change Change { get; set; }

        public int ChangeId { get; set; }

        [Key]
        public int DocumentId { get; set; }
}

public class Media
{
        [ForeignKey(nameof(ChangeId))]
        public Change Change { get; set; }

        public int ChangeId { get; set; }

        [Key]
        public int MediaId { get; set; }
}
```

And here is how we create these entities:

```c#
Document document = new Document() {
    // ...
};

Change change = new Change() {
    // ...
};

document.change = change;

await _context.Documents.AddAsync(document);

await _context.SaveChangesAsync();
```

Now, we noticed that with these relationships and attributes alone (no further configuration in `OnModelCreating`), only the FK constraint on each table other than Changes was being generated. Therefore, when we would add an entity (a `Document`, for instance), the `ChangeId` column for the record in the Documents table was set, but the `DocumentId` column for the record in the Changes was `NULL`. Upon inspecting the migration and the constraints in the database, the FK constraints are indeed missing on the Changes table.

We discovered that we could get the FK constraints generated for Changes by doing the following in `OnModelCreating`:

```c#
            modelBuilder.Entity<Change>()
                .HasOne(c => c.Document)
                .WithOne(d => d.Change)
                .IsRequired(false);

            modelBuilder.Entity<Change>()
                .HasOne(c => c.Media)
                .WithOne(m => m.Change)
                .IsRequired(false);
```

With that change, the opposite happened: the FK constraints for the relationships to each entity from the Changes table were NOT being generated, but the FK constraint from each of the other entities' tables to the Changes table WAS now being generated. Therefore, now when entities are being saved (again, a `Document`, for instance), the `DocumentId` column on the Changes table is being set correctly, but the `ChangeId` on the other side of the relationship is now saved as `0`. Again, upon inspecting the migration and the database, the FK constraints on the Changes table now exist, but the FK constraint referencing the Changes table is missing on the others.

I have tried every permutation of `HasOne`, `WithOne`, `HasForeignKey`, `IsRequired(false)`, and `IsRequired(true)` that I can think of, but no matter what, EF will only generate the FK constraint(s) for one side of the relationship.

I do realize now that we should have used `OwnsOne` or whatever other method via attributes to make each entity referencing a `Change` use that FK as its own PK as well, since that seems to be the norm from what I'm seeing for this kind of relationship, but unfortunately we just realized this is a problem and the hoops we would (or will have to) jump through are cumbersome in order to get the relationship straightened out.

However, I do believe this is a bug - we are telling EF that each entity references the other and what the `[ForeignKey]` is on both sides of the relationship, so why are constraints only being generated for one side of the relationship?

### Further technical details
EF Core version: 2.1.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio Professional 2017 15.9.7

**Comment by IRCraziestTaxi** [comment]
Update: I tried adding `[InverseProperty]` to all relationships to see if that would force EF to honor the relationship in both directions, like so:

```c#
public class Change
{
        [Key]
        public int ChangeId { get; set; }

        [ForeignKey(nameof(DocumentId))]
        [InverseProperty(nameof(Entities.Document.Change))]
        public Document Document { get; set; }

        public int? DocumentId { get; set; }

        [ForeignKey(nameof(MediaId))]
        [InverseProperty(nameof(Entities.Media.Change))]
        public Media Media { get; set; }

        public int? MediaId { get; set; }
}

public class Document
{
        [ForeignKey(nameof(ChangeId))]
        [InverseProperty(nameof(Entities.Change.Document))]
        public Change Change { get; set; }

        public int ChangeId { get; set; }

        [Key]
        public int DocumentId { get; set; }
}

public class Media
{
        [ForeignKey(nameof(ChangeId))]
        [InverseProperty(nameof(Entities.Change.Media))]
        public Change Change { get; set; }

        public int ChangeId { get; set; }

        [Key]
        public int MediaId { get; set; }
}
```

But now, upon attempting to generate a migration, I get the following error:

```
Invalid relationship has been specified using InversePropertyAttribute and ForeignKeyAttribute. The navigation 'Change' in entity type 'Document' and the navigation 'Document' in entity type 'Change' are related by InversePropertyAttribute but the ForeignKeyAttribute specified for both navigations have different values.
```

Of course they have different values... they are different entities that reference each other. What am I missing about what EF wants?

**Comment by smitpatel** [comment]
Can you elaborate what you do mean by relationship in both direction? Relationalships can contain navigations in both direction but foreign key property exists only on 1 entity. And corresponding table in database gets constraints. Having FK on both side for single relationship is incorrect modeling.

**Comment by IRCraziestTaxi** [comment]
@smitpatel Thank you for your response. Our models were designed so that different entity types each must reference a unique instance of this `Change` entity, so that `Change` entity referenced by an entity references that entity alone; the relationships except for the entity corresponding with the `Change` are null. Therefore, the one-to-one relationship FROM the other entity TO the `Change` is required while all relationships FROM the `Change` TO the other entities are optional (since the `Change` only corresponds to one type of entity, not multiple entities). A `Change` cannot be referenced by more than one entity as it is unique to that entity.

Therefore, we thought it logical that each side of the relationship should be able to reference the other so that we can query `Change`s with respect to the one entity the reference and can also query other entities with respect to the `Change` they reference. Since the `Change` can be referenced by different types of entities, it is not an everyday one-to-one relationship.

Sorry if all that was confusing. I tried to fully explain the scenario while being as concise as possible.

**Comment by ajcvickers** [comment]
@IRCraziestTaxi In EF, as in a relational database, each relationship is based on a single foreign key constraint. Consider these entity instances:
```C#
```
The `DocumentId` FK on `change1` indicates it is related to `document2`. But the `ChangeId` on `document2` indicates that it is related to `change3`. So these two FKs cannot represent the same relationship in a normal EF or database model.

Now that is not to say you can't have business logic to keep these FKs in sync, but this is something your code needs to do; it's not inherent in either the EF or database relationship. So, in an EF model like this:
```C#
public class Change
{
    public int ChangeId { get; set; }

    [ForeignKey(nameof(DocumentId))]
    public Document Document { get; set; }
    public int? DocumentId { get; set; }

    [ForeignKey(nameof(MediaId))] 
    public Media Media { get; set; }
    public int? MediaId { get; set; }
}

public class Document
{
    public int DocumentId { get; set; }

    [ForeignKey(nameof(ChangeId))]
    public Change Change { get; set; }
    public int ChangeId { get; set; }
}
```
there are two relationships between the two entities as described above. Setting the navigation property associated with a given FK will only change that FK. The other FK is a different relationship. You will need to write code to set both and keep them in sync for this type of model to work.

**Comment by IRCraziestTaxi** [comment]
@ajcvickers That makes sense. Thanks for the clarification.

And to both of you... I just realized I wasted your time with this issue. Because that FK constraint does exist on the `Change` side, the entities are in fact related and can be queried based on one another. We don't need to care what the `ChangeId` is on the entities referencing it. The fact that our `ChangeId` on the other entities were being saved as `0` was a huge red herring. I tend to go on wild goose chases like that. Sorry. 😅
